### ISSUE TITLE
Legacy migration from Mongobee can fail when running against a multi-node cluster, due to read-your-own-writes logic in MongoCK

### ISSUE THREAD

**Issue by oliverlockwood** [mongock/mongock#355]
**Describe the bug** 

Performing legacy migration (from Mongobee) causes MongoCK to try to read its own writes, which is not reliably if `readPreference` is `secondaryPreferred` and writes don't wait for full replication.

**To Reproduce**

Steps and conditions to reproduce the behavior:
 - Conditions:
    - Library versions are not believed to matter.  FWIW this was hit in a production deployment where we were using:
       - `spring-boot` version `2.3.4.RELEASE`, which ships with `spring-data-mongodb` version `3.0.4.RELEASE`
       - `mongodb` driver version `4.0.5`
    - Infrastructure: Docker-compose, with restart enabled.  Mongo is running in a 3-node cluster with one instance as primary.  This is important; single-node deployments will not replicate the issue.
    - Mongock version 4.1.14
    - How Mongock is implemented:
       - `@EnableMongock` annotation on the `Application` bean
       - configuration in `application.yml` like:
          ```yml
          mongock:
            legacy-migration.collection-name: dbchangelog
            change-logs-scan-package:
              - com.ooyala.flex.events.persistence.changesets
          ```
    - How Mongock is imported: the BOM is imported into dependencyManagement, and the following two dependencies are added:
      ```xml
      <dependency>
        <groupId>com.github.cloudyrock.mongock</groupId>
        <artifactId>mongock-spring-v5</artifactId>
      </dependency>
      <dependency>
        <groupId>com.github.cloudyrock.mongock</groupId>
        <artifactId>mongodb-springdata-v3-driver</artifactId>
      </dependency>
      ```
    - In general anything in your context that could be affecting:
      - Our `MongoClient` bean is instantiated with a `readPreference` of `secondaryPreferred`.  We do not set a `writeConcern`, which suggests that we're using the default value of `WriteConcern.ACKNOWLEDGED`
 - Steps:
    1. Run a previous version of the app, which used Mongobee.  A number of migrations (8) were already present in the `dbchangelog` collection.
    2. Upgrade to and run the new version of the app, which uses MongoCK.  The following behaviour **can** occur - it's obviously a window condition, so doesn't happen every time.  Observe the logs (see below)

**Expected behavior**

After the legacy migration, MongoCK should be aware of everything it has just legacy-migrated from the old Mongobee changelog, and should not attempt to run a previously-executed migration again.

**Logs**

```
Jun 16, 2021 @ 12:10:26.608 - io.changock.driver.core.lock.DefaultLockManager:    Changock trying to acquire the lock
Jun 16, 2021 @ 12:10:26.624 - io.changock.driver.core.lock.DefaultLockManager:    Changock acquired the lock until: Wed Jun 16 12:13:26 GMT 2021
Jun 16, 2021 @ 12:10:26.629 - io.changock.runner.core.MigrationExecutor:    Changock starting the data migration sequence id[2021-06-16T12:10:26.628-714]...
Jun 16, 2021 @ 12:10:26.664 - io.changock.runner.core.MigrationExecutor:    method[mongockSpringLegacyMigration] with arguments: [com.github.cloudyrock.spring.v5.MongockConfiguration$MongockLegacyMigrationConfig, com.github.cloudyrock.mongock.driver.mongodb.sync.v4.decorator.impl.MongoDataBaseDecoratorImpl, com.github.cloudyrock.mongock.driver.mongodb.springdata.v3.SpringDataMongo3ChangeEntryRepository]
Jun 16, 2021 @ 12:10:26.666 - com.github.cloudyrock.mongock.driver.mongodb.sync.v4.changelogs.LegacyService:    [legacy-migration] - There is no changes count expectation!
Jun 16, 2021 @ 12:10:26.785 - io.changock.runner.core.MigrationExecutor:    APPLIED - ChangeEntry{, "id"="mongock-legacy-migration", "author"="mongock", "class"="MongockSync4LegacyMigrationChangeLog", "method"="mongockSpringLegacyMigration"}
Jun 16, 2021 @ 12:10:26.796 - io.changock.runner.core.MigrationExecutor:    PASSED OVER - ChangeEntry{, "id"="migrationOne", "author"="John Doe", "class"="MongoChangeLog", "method"="migrationOne"}
Jun 16, 2021 @ 12:10:26.798 - io.changock.runner.core.MigrationExecutor:    PASSED OVER - ChangeEntry{, "id"="migrationTwo", "author"="Jane Doe", "class"="MongoChangeLog", "method"="migrationTwo"}
Jun 16, 2021 @ 12:10:26.800 - io.changock.runner.core.MigrationExecutor:    method[migrationThree] with arguments: [com.github.cloudyrock.mongock.driver.mongodb.sync.v4.decorator.impl.MongoDataBaseDecoratorImpl]
Jun 16, 2021 @ 12:11:26.865 - io.changock.runner.core.MigrationExecutor:    FAILED OVER - ChangeEntry{, "id"="migrationThree", "author"="Joshua Doe", "class"="MongoChangeLog", "method"="migrationThree"}
Jun 16, 2021 @ 12:11:26.868 - io.changock.driver.core.lock.DefaultLockManager:    Changock releasing the lock
Jun 16, 2021 @ 12:11:26.872 - io.changock.runner.core.MigrationExecutor:    Changock has finished
Jun 16, 2021 @ 12:11:26.872 - io.changock.driver.core.lock.DefaultLockManager:    Changock released the lock
```

**Screenshots**

No screenshot is relevant here.

**Additional context**

After the failure, when the app starts up again MongoCK notices that `migrationThree` (and all the other previously-performed migrations) were already done, and the app starts up just fine.  However, in addition to the (legacy-migrated) `EXECUTED` entries in the new MongoCK change log, there's a `FAILED` entry for `migrationThree` still in there.

After some analysis and code read of MongoCK, it became obvious that this is a window condition.  Since our Mongo client write concern is not to wait until all replicas are updated, and our `readPreference` is `secondaryPreferred`, clearly the "normal" migration logic in MongoCK looked for the existence of `migrationThree` from a secondary node to which replication had not yet completed (after the "legacy migration" wrote the data there).

**Comment by dieppa** [comment]
Thanks @oliverlockwood ,

we are currently a bit busy with the some clients, but we'll take a look and will get back to you as soon as we can ;)

**Comment by oliverlockwood** [comment]
Incidentally, I was curious as to who your clients may be, and so I went to your organisation page https://github.com/cloudyrock.  This has a link to http://cloudyrock.io/ ... which is apparently parked by GoDaddy.  You might want to update the link on your Github organisation - or claim back your domain!

**Comment by dieppa** [comment]
Hi again,

I am glad you ask..that shows some interest in Mongock 😃 

Although Cloudyrock will be still the company behind, we are moving everything to mongck.io, where there is currently the Mongock's documentation.

Regarding your question about our clients. We can't disclose our client names yet, but we will soon share these through some of our public endorsements through our new website. But they are some early users of the support model and professional library will be releasing as part of the version 5.

We'll try to give you a response on your issue as soon as we can 😉

**Comment by dieppa** [comment]
Hello,

after having a quick look, from my understanding, this would happen to a normal migration. It's not exclusively for the legacy migration, right? If not, please tell me why because I would had take something wrong.


Assuming the statement above is correct, I understand you don't want to have consistency in your application.

Is there any problem in having consistency only in Mongock?
Please see this [issue](https://github.com/cloudyrock/mongock/issues/307). It's still not documented, but you can set the MongoDB write/read concern and preference only for Mongock.


That should fix your issue. Please keep us updated 😃

**Comment by oliverlockwood** [comment]
Oops, fat fingers, didn't mean to close that.

Thanks for the pointer on how to configure MongoCK with different settings from our application's read preference / write concern. We will look into that and get back to you.

Regarding the issue, I believe that it is specific to the legacy migration case, as follows.

**On a normal migration**, each migration is checked in turn, with a look-up in the MongoCK changelog collection to see if it's been previously run.
- if previously run, the lookup will succeed (in replication terms, it's been ages since the previous migration operation, so all replicas are up to date) and so the individual migration will (correctly) not be run
- otherwise, the lookup will fail, and the migration will (correctly) run.

**On a legacy migration**, the legacy Mongobee changelog collection is read, and its contents written to the MongoCK changelog collection. Then, the same steps as normal (above) are immediately run.
- in this case, it has **not** been ages, in replication terms, since the previously-run migrations were written to the MongoCK changelog collection.
- therefore, when each migration is checked in turn, it is possible for the lookup to fail and the migration to be run even though the migration had already previously been performed.

Hopefully the above is clear; let me know and if not I'll dig out code references to be more specific.

**Comment by dieppa** [comment]
Thanks, I am not able to look at that now. but I will as soon as I can and I will get back to you.

Cheers

**Comment by dieppa** [comment]
Sorry, @oliverlockwood , we normally fix things very soon, but it seems we forgot about this one. Does it still make sense?

Cheers

**Comment by oliverlockwood** [comment]
Hi @dieppa, no worries.

Urgency is not high from our particular perspective because the impact we experienced was not critical, and most deployments we have now already done the Mongobee --> MongoCK migration, so will no longer be at risk.  But for some users I could imagine that re-running a migration could cause potentially serious problems, so if it were my project I'd want to fix it 😄 

This was definitely an issue on MongoCK 4.1.14, as noted originally.

From reading the latest documentation  https://mongock.io/v5/driver/mongodb/index.html#mongodb-common-configuration it looks like you now set both `readConcern` & `writeConcern`to `majority`, which from my understand of the meaning of these fields, should be sufficient to prevent this window condition from happening.

So I guess we can close this as "probably fixed in 5.x" and consider re-opening if it gets hit again.  How does that sound?

**Comment by dieppa** [comment]
Thanks @oliverlockwood ,

We will double check this is the case and it's indeed fixed in Mongock 5.0.x.

Thanks again!

**Comment by dieppa** [comment]
It was indeed a problem on that. We fixed it. By default it works as expected, however, if the user decided to relax the read/write configuration, it will affect. We must leave this window open as this high consistent configuration might be problematic in some environments, dev for example.

Thanks
